testsrc/org/mozilla/javascript/xmlimpl/XmlNonResettableDocumentBuilderTest.java

```diff
@@ -0,0 +1,43 @@
+package org.mozilla.javascript.xmlimpl;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mozilla.javascript.Context;
+import org.mozilla.javascript.ContextFactory;
+import org.mozilla.javascript.Scriptable;
+
+public class XmlNonResettableDocumentBuilderTest {
+    private static final String XML_PROPERTY = "javax.xml.parsers.DocumentBuilderFactory";
+    private final String originalDocumentBuilderFactory = System.getProperty(XML_PROPERTY);
+
+    @Before
+    public void setUp() {
+        System.setProperty(XML_PROPERTY, NonResettableDocumentBuilderFactory.class.getName());
+    }
+
+    @After
+    public void tearDown() {
+        if (originalDocumentBuilderFactory == null) {
+            System.clearProperty(XML_PROPERTY);
+        } else {
+            System.setProperty(XML_PROPERTY, originalDocumentBuilderFactory);
+        }
+    }
+
+    @Test
+    public void testNonResettableDocumentBuilder() {
+        Context cx = new ContextFactory().enterContext();
+        try {
+            Scriptable scope = cx.initStandardObjects();
+            Object result = cx.evaluateString(scope,
```

Comment:
I've been reading but I don't understand how this test case tests the resetting of the document builder. Don't we need to at least run the XML processor twice so that we can observe the attempt to reset the builder and then cache the result? Or is there something about how we configure the builder in the XML processing code that will fail if it doesn't happen here? In other words, is this a test that actually failed before this change was added?

Reply:
Yes, this is a real test: if you comment out the workaround in XmlProcessor.returnDocumentBuilderToPool - the test will fail.

The thing is both XmlProcessor.getDocumentBuilderFromPool() and XmlProcessor.returnDocumentBuilderToPool() are called twice during the test:
1. When XML library is initialized, for empty XML string
2. When the actual XML string from the test is read

Reply:
Thanks -- didn't realize that it's invoked multiple times anyway.